lib/spanner/config.ex

```diff
@@ -40,19 +49,44 @@ defmodule Spanner.Config do
   def fixup_rules(config),
     do: config
 
-  @doc "Validate bundle configs"
-  def validate(config) do
-    case Spanner.Config.SyntaxValidator.validate(config) do
+  @doc """
+  Validates bundle configs. Returns {:ok, config} | {:error, errors, warnings} |
+  {:warning, config, warnings}
+  """
+  @spec validate(Map.t) ::
+    {:ok, Map.t} | {:error, List.t, List.t} | {:warning, Map.t, List.t}
+  def validate(%{"cog_bundle_version" => @current_config_version}=config) do
+    case SyntaxValidator.validate(config) do
       :ok ->
         config = fixup_rules(config)
-        case Spanner.Config.SemanticValidator.validate(config) do
+        case SemanticValidator.validate(config) do
           :ok ->
             {:ok, config}
-          error ->
-            error
+          {:error, errors} ->
+            {:error, errors, []}
+        end
+      {:error, errors} ->
+        {:error, errors, []}
+    end
+  end
+  # Gets triggered when we have an old bundle config to validate
+  # Upgrader will return an error if the bundle is not upgradable
+  def validate(config) do
```

Comment:
Why not put the N-1 version number explicitly here? Makes the intent much more obvious, imo.